File: cmangos-cms-api.client/electron/services/ServerControlService.ts
import { spawn } from 'child_process';
import { createWriteStream, existsSync, promises as fs } from 'fs';
import { dirname, join, normalize } from 'path';
import type { ServerProcessName, ServerProcessStatus } from '../types/config.types.js';

interface ServerPaths {
  realmdPath: string;
  mangosdPath: string;
}

interface ServerStartOptions {
  showConsole?: boolean;
}

interface ProcessInfo {
  ProcessId?: number;
  ExecutablePath?: string;
}

interface ServerLogOutput {
  stdout: string;
  stderr: string;
}

interface ServerLogsSnapshot {
  realmd: ServerLogOutput;
  mangosd: ServerLogOutput;
}

export class ServerControlService {
  private startedByApp = new Map<ServerProcessName, number>();

  private ensureWindows(): void {
    if (process.platform !== 'win32') {
      throw new Error('Server control is only supported on Windows right now.');
    }
  }

  private normalizePath(value: string): string {
    return normalize(value).toLowerCase();
  }

  private resolveExecutablePath(inputPath: string, exeName: string): string {
    const trimmed = inputPath.trim();
    if (!trimmed) {
      throw new Error(`${exeName} path is required`);
    }

    const lower = trimmed.toLowerCase();
    if (lower.endsWith('.exe') && existsSync(trimmed)) {
      return trimmed;
    }

    const candidate = join(trimmed, exeName);
    if (existsSync(candidate)) {
      return candidate;
    }

    throw new Error(`Could not find ${exeName} at the specified path.`);
  }

  private runPowerShell(command: string): Promise<string> {
    return new Promise((resolve, reject) => {
      const child = spawn('powershell.exe', ['-NoProfile', '-Command', command], {
        windowsHide: true
      });

      let stdout = '';
      let stderr = '';

      child.stdout?.on('data', (data) => {
        stdout += data.toString();
      });

      child.stderr?.on('data', (data) => {
        stderr += data.toString();
      });

      child.on('error', (error) => {
        reject(error);
      });

      child.on('close', (code) => {
        if (code !== 0 && stderr.trim()) {
          reject(new Error(stderr.trim()));
          return;
        }
        resolve(stdout.trim());
      });
    });
  }

  private async getProcessInfos(exeName: string): Promise<ProcessInfo[]> {
    const command = `Get-CimInstance Win32_Process -Filter "Name='${exeName}'" | Select-Object ProcessId, ExecutablePath | ConvertTo-Json -Compress`;
    const output = await this.runPowerShell(command);

    if (!output || output === 'null') {
      return [];
    }

    const parsed = JSON.parse(output) as ProcessInfo | ProcessInfo[];
    return Array.isArray(parsed) ? parsed : [parsed];
  }

  private async findMatchingProcess(executablePath: string): Promise<ProcessInfo | null> {
    const exeName = normalize(executablePath).split('\\').pop() || '';
    const processes = await this.getProcessInfos(exeName);
    const normalizedTarget = this.normalizePath(executablePath);

    const match = processes.find((process) => {
      if (!process.ExecutablePath) {
        return false;
      }
      return this.normalizePath(process.ExecutablePath) === normalizedTarget;
    });

    return match || null;
  }

  private async killProcess(pid: number): Promise<void> {
    await new Promise<void>((resolve, reject) => {
      const child = spawn('taskkill', ['/PID', String(pid), '/T', '/F'], {
        windowsHide: true
      });

      child.on('error', reject);
      child.on('close', () => resolve());
    });
  }

  private async readLogTail(filePath: string, maxBytes = 20000): Promise<string> {
    try {
      const content = await fs.readFile(filePath, 'utf8');
      if (content.length <= maxBytes) {
        return content;
      }
      return content.slice(-maxBytes);
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
        return '';
      }
      throw error;
    }
  }

  private async getLogOutput(name: ServerProcessName, inputPath: string, exeName: string): Promise<ServerLogOutput> {
    const executablePath = this.resolveExecutablePath(inputPath, exeName);
    const workingDir = dirname(executablePath);
    const stdoutLogPath = join(workingDir, `${name}.stdout.log`);
    const stderrLogPath = join(workingDir, `${name}.stderr.log`);

    const [stdout, stderr] = await Promise.all([
      this.readLogTail(stdoutLogPath),
      this.readLogTail(stderrLogPath)
    ]);

    return { stdout, stderr };
  }

  private async getServerStatus(name: ServerProcessName, inputPath: string, exeName: string): Promise<ServerProcessStatus> {
    try {
      const executablePath = this.resolveExecutablePath(inputPath, exeName);
      const match = await this.findMatchingProcess(executablePath);
      if (match?.ProcessId) {
        return {
          name,
          status: 'running',
          pid: match.ProcessId,
          executablePath,
          startedByApp: this.startedByApp.get(name) === match.ProcessId
        };
      }

      return {
        name,
        status: 'stopped',
        executablePath
      };
    } catch (error) {
      return {
        name,
        status: 'unknown',
        error: (error as Error).message
      };
    }
  }

  async validatePaths(paths: ServerPaths): Promise<ServerPaths> {
    this.ensureWindows();

    return {
      realmdPath: this.resolveExecutablePath(paths.realmdPath, 'realmd.exe'),
      mangosdPath: this.resolveExecutablePath(paths.mangosdPath, 'mangosd.exe')
    };
  }

  async getStatus(paths: ServerPaths): Promise<ServerProcessStatus[]> {
    this.ensureWindows();

    const [realmdStatus, mangosdStatus] = await Promise.all([
      this.getServerStatus('realmd', paths.realmdPath, 'realmd.exe'),
      this.getServerStatus('mangosd', paths.mangosdPath, 'mangosd.exe')
    ]);

    return [realmdStatus, mangosdStatus];
  }

  async getLogs(paths: ServerPaths): Promise<ServerLogsSnapshot> {
    this.ensureWindows();

    const [realmd, mangosd] = await Promise.all([
      this.getLogOutput('realmd', paths.realmdPath, 'realmd.exe'),
      this.getLogOutput('mangosd', paths.mangosdPath, 'mangosd.exe')
    ]);

    return { realmd, mangosd };
  }

  private async startServer(
    name: ServerProcessName,
    inputPath: string,
    exeName: string,
    args: string[],
    options: ServerStartOptions
  ): Promise<void> {
    const executablePath = this.resolveExecutablePath(inputPath, exeName);
    const running = await this.findMatchingProcess(executablePath);
    if (running?.ProcessId) {
      return;
    }

    const workingDir = dirname(executablePath);

    let child;

    if (options.showConsole) {
      child = spawn('cmd.exe', ['/c', 'start', '""', executablePath, ...args], {
        cwd: workingDir,
        windowsHide: false
      });
    } else {
      const stdoutLogPath = join(workingDir, `${name}.stdout.log`);
      const stderrLogPath = join(workingDir, `${name}.stderr.log`);
      const stdoutStream = createWriteStream(stdoutLogPath, { flags: 'a' });
      const stderrStream = createWriteStream(stderrLogPath, { flags: 'a' });

      child = spawn(executablePath, args, {
        cwd: workingDir,
        detached: true,
        stdio: ['ignore', 'pipe', 'pipe'],
        windowsHide: true
      });

      child.stdout?.pipe(stdoutStream);
      child.stderr?.pipe(stderrStream);

      child.on('exit', (code, signal) => {
        const message = `[ServerControl] ${exeName} exited (code=${code ?? 'null'}, signal=${signal ?? 'null'}). Logs: ${stdoutLogPath}, ${stderrLogPath}`;
        console.warn(message);
      });
    }

    child.on('error', (error) => {
      console.error(`[ServerControl] Failed to spawn ${exeName}:`, error);
    });

    if (!child.pid) {
      throw new Error(`Failed to start ${exeName}`);
    }

    child.unref();
    this.startedByApp.set(name, child.pid);
  }

  async start(paths: ServerPaths, options: ServerStartOptions = {}): Promise<ServerProcessStatus[]> {
    this.ensureWindows();

    await this.startServer('realmd', paths.realmdPath, 'realmd.exe', ['-c', 'realmd.conf'], options);

    const mangosExecutablePath = this.resolveExecutablePath(paths.mangosdPath, 'mangosd.exe');
    const mangosDir = dirname(mangosExecutablePath);
    const mangosArgs = ['-c', 'mangosd.conf'];
    const ahbotConfigPath = join(mangosDir, 'ahbot.conf');
    if (existsSync(ahbotConfigPath)) {
      mangosArgs.push('-a', 'ahbot.conf');
    }

    await this.startServer('mangosd', paths.mangosdPath, 'mangosd.exe', mangosArgs, options);

    return this.getStatus(paths);
  }

  private async stopServer(name: ServerProcessName, inputPath: string, exeName: string): Promise<void> {
    const executablePath = this.resolveExecutablePath(inputPath, exeName);
    const processes = await this.getProcessInfos(exeName);
    const normalizedTarget = this.normalizePath(executablePath);

    const matches = processes.filter((process) =>
      process.ExecutablePath && this.normalizePath(process.ExecutablePath) === normalizedTarget
    );

    for (const processInfo of matches) {
      if (processInfo.ProcessId) {
        await this.killProcess(processInfo.ProcessId);
      }
    }

    this.startedByApp.delete(name);
  }

  async stop(paths: ServerPaths): Promise<ServerProcessStatus[]> {
    this.ensureWindows();

    await this.stopServer('realmd', paths.realmdPath, 'realmd.exe');
    await this.stopServer('mangosd', paths.mangosdPath, 'mangosd.exe');

    return this.getStatus(paths);
  }

  async restart(paths: ServerPaths, options: ServerStartOptions = {}): Promise<ServerProcessStatus[]> {
    this.ensureWindows();

    await this.stop(paths);
    await this.start(paths, options);

    return this.getStatus(paths);
  }
}

File: CMAnGOS-CMS-API.Server/Services/ServerControlService.cs
using System.Collections.Concurrent;
using System.Diagnostics;
using System.Management;
using System.Runtime.Versioning;
using System.Text;
using CMAnGOS_CMS_API.Server.Models.Dto;

namespace CMAnGOS_CMS_API.Server.Services
{
    [SupportedOSPlatform("windows")]
    public sealed class ServerControlService : IServerControlService
    {
        private readonly ConcurrentDictionary<string, int> _startedByApp = new(StringComparer.OrdinalIgnoreCase);

        private static void EnsureWindows()
        {
            if (!OperatingSystem.IsWindows())
            {
                throw new InvalidOperationException("Server control is only supported on Windows.");
            }
        }

        private static string NormalizePath(string value)
        {
            return Path.GetFullPath(value).Trim().ToLowerInvariant();
        }

        private static string ResolveExecutablePath(string inputPath, string exeName)
        {
            var trimmed = inputPath?.Trim() ?? string.Empty;
            if (string.IsNullOrWhiteSpace(trimmed))
            {
                throw new InvalidOperationException($"{exeName} path is required.");
            }

            if (trimmed.EndsWith(".exe", StringComparison.OrdinalIgnoreCase) && File.Exists(trimmed))
            {
                return trimmed;
            }

            var candidate = Path.Combine(trimmed, exeName);
            if (File.Exists(candidate))
            {
                return candidate;
            }

            throw new InvalidOperationException($"Could not find {exeName} at the specified path.");
        }

        [SupportedOSPlatform("windows")]
        private static IReadOnlyList<(int? ProcessId, string? ExecutablePath)> GetProcessInfos(string exeName)
        {
            if (!OperatingSystem.IsWindows())
            {
                return Array.Empty<(int?, string?)>();
            }

            var results = new List<(int?, string?)>();
            using var searcher = new ManagementObjectSearcher(
                $"SELECT ProcessId, ExecutablePath FROM Win32_Process WHERE Name='{exeName}'");

            foreach (var item in searcher.Get().OfType<ManagementObject>())
            {
                var pidValue = item["ProcessId"];
                var pathValue = item["ExecutablePath"];
                var pid = pidValue != null ? Convert.ToInt32(pidValue) : (int?)null;
                var path = pathValue?.ToString();
                results.Add((pid, path));
            }

            return results;
        }

        private static (int? ProcessId, string? ExecutablePath)? FindMatchingProcess(string executablePath)
        {
            var exeName = Path.GetFileName(executablePath);
            var target = NormalizePath(executablePath);

            var matches = GetProcessInfos(exeName)
                .Where(info => !string.IsNullOrWhiteSpace(info.ExecutablePath))
                .FirstOrDefault(info => NormalizePath(info.ExecutablePath!) == target);

            return matches.ExecutablePath == null ? null : matches;
        }

        private static async Task KillProcessAsync(int pid, CancellationToken cancellationToken)
        {
            try
            {
                var process = Process.GetProcessById(pid);
                process.Kill(entireProcessTree: true);
                await process.WaitForExitAsync(cancellationToken);
            }
            catch (ArgumentException)
            {
                // Process already exited.
            }
        }

        private static async Task<string> ReadLogTailAsync(string filePath, int maxBytes, CancellationToken cancellationToken)
        {
            if (!File.Exists(filePath))
            {
                return string.Empty;
            }

            await using var stream = new FileStream(filePath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
            if (stream.Length <= maxBytes)
            {
                using var reader = new StreamReader(stream, Encoding.UTF8);
                return await reader.ReadToEndAsync(cancellationToken);
            }

            stream.Seek(-maxBytes, SeekOrigin.End);
            using var tailReader = new StreamReader(stream, Encoding.UTF8);
            return await tailReader.ReadToEndAsync(cancellationToken);
        }

        private static string BuildArgumentString(IEnumerable<string> args)
        {
            return string.Join(" ", args.Select(QuoteArgument));
        }

        private static string QuoteArgument(string value)
        {
            return value.Contains(' ') ? $"\"{value}\"" : value;
        }

        private static async Task PumpToFileAsync(StreamReader reader, string filePath, CancellationToken cancellationToken)
        {
            await using var stream = new FileStream(filePath, FileMode.Append, FileAccess.Write, FileShare.ReadWrite);
            await using var writer = new StreamWriter(stream) { AutoFlush = true };

            while (!reader.EndOfStream && !cancellationToken.IsCancellationRequested)
            {
                var line = await reader.ReadLineAsync(cancellationToken);
                if (line == null)
                {
                    break;
                }

                await writer.WriteLineAsync(line);
            }
        }

        private static async Task<ServerLogOutputDto> GetLogOutputAsync(string inputPath, string exeName, string logPrefix, CancellationToken cancellationToken)
        {
            var executablePath = ResolveExecutablePath(inputPath, exeName);
            var workingDir = Path.GetDirectoryName(executablePath) ?? string.Empty;
            var stdoutLogPath = Path.Combine(workingDir, $"{logPrefix}.stdout.log");
            var stderrLogPath = Path.Combine(workingDir, $"{logPrefix}.stderr.log");

            var stdout = await ReadLogTailAsync(stdoutLogPath, 20000, cancellationToken);
            var stderr = await ReadLogTailAsync(stderrLogPath, 20000, cancellationToken);

            return new ServerLogOutputDto { Stdout = stdout, Stderr = stderr };
        }

        private static IReadOnlyList<string> BuildMangosArgs(string mangosdPath)
        {
            var mangosExecutablePath = ResolveExecutablePath(mangosdPath, "mangosd.exe");
            var mangosDir = Path.GetDirectoryName(mangosExecutablePath) ?? string.Empty;
            var mangosArgs = new List<string> { "-c", "mangosd.conf" };
            var ahbotConfigPath = Path.Combine(mangosDir, "ahbot.conf");
            if (File.Exists(ahbotConfigPath))
            {
                mangosArgs.Add("-a");
                mangosArgs.Add("ahbot.conf");
            }

            return mangosArgs;
        }

        private ServerProcessStatusDto GetServerStatus(string name, string inputPath, string exeName)
        {
            try
            {
                var executablePath = ResolveExecutablePath(inputPath, exeName);
                var match = FindMatchingProcess(executablePath);
                if (match is { ProcessId: not null } found)
                {
                    return new ServerProcessStatusDto
                    {
                        Name = name,
                        Status = "running",
                        Pid = found.ProcessId,
                        ExecutablePath = executablePath,
                        StartedByApp = _startedByApp.TryGetValue(name, out var trackedPid) && found.ProcessId == trackedPid
                    };
                }

                return new ServerProcessStatusDto
                {
                    Name = name,
                    Status = "stopped",
                    ExecutablePath = executablePath
                };
            }
            catch (Exception ex)
            {
                return new ServerProcessStatusDto
                {
                    Name = name,
                    Status = "unknown",
                    Error = ex.Message
                };
            }
        }

        public Task<ServerPathsRequest> ValidatePathsAsync(ServerPathsRequest request, CancellationToken cancellationToken = default)
        {
            EnsureWindows();
            _ = cancellationToken;

            return Task.FromResult(new ServerPathsRequest
            {
                RealmdPath = ResolveExecutablePath(request.RealmdPath, "realmd.exe"),
                MangosdPath = ResolveExecutablePath(request.MangosdPath, "mangosd.exe")
            });
        }

        public Task<IReadOnlyList<ServerProcessStatusDto>> GetStatusAsync(ServerPathsRequest request, CancellationToken cancellationToken = default)
        {
            EnsureWindows();
            _ = cancellationToken;

            var statuses = new List<ServerProcessStatusDto>
            {
                GetServerStatus("realmd", request.RealmdPath, "realmd.exe"),
                GetServerStatus("mangosd", request.MangosdPath, "mangosd.exe")
            };

            return Task.FromResult<IReadOnlyList<ServerProcessStatusDto>>(statuses);
        }

        public Task<ServerProcessStatusDto> GetRealmdStatusAsync(RealmdPathRequest request, CancellationToken cancellationToken = default)
        {
            EnsureWindows();
            _ = cancellationToken;
            return Task.FromResult(GetServerStatus("realmd", request.RealmdPath, "realmd.exe"));
        }

        public Task<ServerProcessStatusDto> GetMangosdStatusAsync(MangosdPathRequest request, CancellationToken cancellationToken = default)
        {
            EnsureWindows();
            _ = cancellationToken;
            return Task.FromResult(GetServerStatus("mangosd", request.MangosdPath, "mangosd.exe"));
        }

        public async Task<IReadOnlyList<ServerProcessStatusDto>> StartAsync(ServerPathsRequest request, CancellationToken cancellationToken = default)
        {
            EnsureWindows();

            await StartRealmdAsync(new RealmdPathRequest
            {
                RealmdPath = request.RealmdPath,
                ShowConsole = request.ShowConsole
            }, cancellationToken);

            await StartMangosdAsync(new MangosdPathRequest
            {
                MangosdPath = request.MangosdPath,
                ShowConsole = request.ShowConsole
            }, cancellationToken);

            return await GetStatusAsync(request, cancellationToken);
        }

        public async Task<ServerProcessStatusDto> StartRealmdAsync(RealmdPathRequest request, CancellationToken cancellationToken = default)
        {
            EnsureWindows();
            await StartServerAsync("realmd", request.RealmdPath, "realmd.exe", new[] { "-c", "realmd.conf" }, request.ShowConsole, cancellationToken);
            return await GetRealmdStatusAsync(request, cancellationToken);
        }

        public async Task<ServerProcessStatusDto> StartMangosdAsync(MangosdPathRequest request, CancellationToken cancellationToken = default)
        {
            EnsureWindows();
            var args = BuildMangosArgs(request.MangosdPath);
            await StartServerAsync("mangosd", request.MangosdPath, "mangosd.exe", args, request.ShowConsole, cancellationToken);
            return await GetMangosdStatusAsync(request, cancellationToken);
        }

        public async Task<IReadOnlyList<ServerProcessStatusDto>> StopAsync(ServerPathsRequest request, CancellationToken cancellationToken = default)
        {
            EnsureWindows();

            await StopRealmdAsync(new RealmdPathRequest { RealmdPath = request.RealmdPath }, cancellationToken);
            await StopMangosdAsync(new MangosdPathRequest { MangosdPath = request.MangosdPath }, cancellationToken);

            return await GetStatusAsync(request, cancellationToken);
        }

        public async Task<ServerProcessStatusDto> StopRealmdAsync(RealmdPathRequest request, CancellationToken cancellationToken = default)
        {
            EnsureWindows();
            await StopServerAsync("realmd", request.RealmdPath, "realmd.exe", cancellationToken);
            return await GetRealmdStatusAsync(request, cancellationToken);
        }

        public async Task<ServerProcessStatusDto> StopMangosdAsync(MangosdPathRequest request, CancellationToken cancellationToken = default)
        {
            EnsureWindows();
            await StopServerAsync("mangosd", request.MangosdPath, "mangosd.exe", cancellationToken);
            return await GetMangosdStatusAsync(request, cancellationToken);
        }

        public async Task<IReadOnlyList<ServerProcessStatusDto>> RestartAsync(ServerPathsRequest request, CancellationToken cancellationToken = default)
        {
            EnsureWindows();

            await RestartRealmdAsync(new RealmdPathRequest
            {
                RealmdPath = request.RealmdPath,
                ShowConsole = request.ShowConsole
            }, cancellationToken);
            await RestartMangosdAsync(new MangosdPathRequest
            {
                MangosdPath = request.MangosdPath,
                ShowConsole = request.ShowConsole
            }, cancellationToken);

            return await GetStatusAsync(request, cancellationToken);
        }

        public async Task<ServerProcessStatusDto> RestartRealmdAsync(RealmdPathRequest request, CancellationToken cancellationToken = default)
        {
            EnsureWindows();
            await StopRealmdAsync(request, cancellationToken);
            await StartRealmdAsync(request, cancellationToken);
            return await GetRealmdStatusAsync(request, cancellationToken);
        }

        public async Task<ServerProcessStatusDto> RestartMangosdAsync(MangosdPathRequest request, CancellationToken cancellationToken = default)
        {
            EnsureWindows();
            await StopMangosdAsync(request, cancellationToken);
            await StartMangosdAsync(request, cancellationToken);
            return await GetMangosdStatusAsync(request, cancellationToken);
        }

        public async Task<ServerLogsSnapshotDto> GetLogsAsync(ServerPathsRequest request, CancellationToken cancellationToken = default)
        {
            EnsureWindows();

            var realmd = await GetLogOutputAsync(request.RealmdPath, "realmd.exe", "realmd", cancellationToken);
            var mangosd = await GetLogOutputAsync(request.MangosdPath, "mangosd.exe", "mangosd", cancellationToken);

            return new ServerLogsSnapshotDto { Realmd = realmd, Mangosd = mangosd };
        }

        public async Task<ServerLogOutputDto> GetRealmdLogsAsync(RealmdPathRequest request, CancellationToken cancellationToken = default)
        {
            EnsureWindows();
            return await GetLogOutputAsync(request.RealmdPath, "realmd.exe", "realmd", cancellationToken);
        }

        public async Task<ServerLogOutputDto> GetMangosdLogsAsync(MangosdPathRequest request, CancellationToken cancellationToken = default)
        {
            EnsureWindows();
            return await GetLogOutputAsync(request.MangosdPath, "mangosd.exe", "mangosd", cancellationToken);
        }

        private Task StartServerAsync(
            string name,
            string inputPath,
            string exeName,
            IEnumerable<string> args,
            bool? showConsole,
            CancellationToken cancellationToken)
        {
            var executablePath = ResolveExecutablePath(inputPath, exeName);
            var running = FindMatchingProcess(executablePath);
            if (running?.ProcessId != null)
            {
                return Task.CompletedTask;
            }

            var workingDir = Path.GetDirectoryName(executablePath) ?? string.Empty;

            if (showConsole == true)
            {
                var commandArgs = $"/c start \"\" {QuoteArgument(executablePath)} {BuildArgumentString(args)}";
                var startInfo = new ProcessStartInfo
                {
                    FileName = "cmd.exe",
                    Arguments = commandArgs,
                    WorkingDirectory = workingDir,
                    UseShellExecute = true,
                    CreateNoWindow = false
                };

                Process.Start(startInfo);
                return Task.CompletedTask;
            }

            var stdoutLogPath = Path.Combine(workingDir, $"{name}.stdout.log");
            var stderrLogPath = Path.Combine(workingDir, $"{name}.stderr.log");

            var process = new Process
            {
                StartInfo = new ProcessStartInfo
                {
                    FileName = executablePath,
                    WorkingDirectory = workingDir,
                    UseShellExecute = false,
                    RedirectStandardOutput = true,
                    RedirectStandardError = true,
                    CreateNoWindow = true
                },
                EnableRaisingEvents = true
            };

            foreach (var arg in args)
            {
                process.StartInfo.ArgumentList.Add(arg);
            }

            process.Start();
            if (process.Id > 0)
            {
                _startedByApp[name] = process.Id;
            }

            _ = Task.Run(() => PumpToFileAsync(process.StandardOutput, stdoutLogPath, cancellationToken), cancellationToken);
            _ = Task.Run(() => PumpToFileAsync(process.StandardError, stderrLogPath, cancellationToken), cancellationToken);

            return Task.CompletedTask;
        }

        private async Task StopServerAsync(string name, string inputPath, string exeName, CancellationToken cancellationToken)
        {
            var executablePath = ResolveExecutablePath(inputPath, exeName);
            var target = NormalizePath(executablePath);
            var processes = GetProcessInfos(exeName)
                .Where(info => info.ExecutablePath != null && NormalizePath(info.ExecutablePath) == target);

            foreach (var processInfo in processes)
            {
                if (processInfo.ProcessId != null)
                {
                    await KillProcessAsync(processInfo.ProcessId.Value, cancellationToken);
                }
            }

            _startedByApp.TryRemove(name, out _);
        }
    }
}
